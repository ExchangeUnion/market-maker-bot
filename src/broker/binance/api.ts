import assert from 'assert';
import axios from 'axios';
import crypto from 'crypto';
import querystring from 'querystring';
import { OrderSide, OrderType, OrderStatus } from '../../enums';
import { Logger } from '../../logger';
import {
  Balance,
  QueryOrderResponse,
  QueryOrderRequest,
  CancelOrderRequest,
  ExchangeAPI,
} from '../api';
import { v4 as uuidv4 } from 'uuid';
import { LimitOrderRequest, StopLimitOrderRequest } from '../exchange';
import { BinanceOrder } from './order';

const API_BASE_URL = 'https://api.binance.com';
const ACCOUNT_URL = `${API_BASE_URL}/api/v3/account`;
const TEST_ORDER_URL = `${API_BASE_URL}/api/v3/order/test`;
const ORDER_URL = `${API_BASE_URL}/api/v3/order`;
const EXCHANGE_INFO_URL = `${API_BASE_URL}/api/v3/exchangeInfo`;
const AVERAGE_PRICE = `${API_BASE_URL}/api/v3/avgPrice`;
const PING_URL = `${API_BASE_URL}/api/v1/ping`;

type QueryOrderResponseBinance = {
  symbol: string;
  status: OrderStatus;
};

export type OrderFill = {
  commission: string;
  commissionAsset: string;
  price: string;
  qty: string;
  tradeId: number;
};

export type OrderResponse = {
  clientOrderId: string;
  cummulativeQuoteQty: string;
  executedQty: string;
  fills: OrderFill[];
  orderId: number;
  orderListId: number;
  origQty: string;
  price: string;
  side: OrderSide;
  status: string;
  symbol: string;
  timeInForce: string;
  transactTime: number;
  type: string;
};

type BaseFilter = {
  filterType: string;
};

type LotSizeFilter = BaseFilter & {
  minQty: string;
  maxQty: string;
  stepSize: string;
};

type MinNotionalFilter = BaseFilter & {
  minNotional: string;
  applyToMarket: boolean;
  avgPriceMins: number;
};

type Filter = LotSizeFilter | MinNotionalFilter;

type TickerSymbol = {
  baseAsset: string;
  quoteAsset: string;
  filters: Filter[];
};

type ExchangeInfoResponse = {
  symbols: TickerSymbol[];
};

type AvgPriceResponse = {
  sym: string;
  price: string;
};

type RawBalance = {
  asset: string;
  free: string;
  locked: string;
};

type AccountInfo = {
  balances: RawBalance[];
};

require('axios-debug-log')({
  request(debug: any, config: any) {
    debug(`Request with ${config.headers['content-type']}`);
  },
  response(debug: any, response: any) {
    debug('Received HTTP response', response);
    debug(`
      Response with ${response.headers['content-type']},
      from ${response.config.url},
    `);
  },
  error(debug: any, error: any) {
    debug('HTTP error', error);
  },
});

const PING_INTERVAL = 180000;
const EXCHANGE_INFO_INTERVAL = 300000;
const UPDATE_AVERAGE_PRICES_INTERVAL = 300000;

class BinanceAPI extends ExchangeAPI {
  private pingInterval: ReturnType<typeof setInterval> | undefined;
  private exchangeInfoInterval: ReturnType<typeof setInterval> | undefined;
  private updateAveragePricesInterval:
    | ReturnType<typeof setInterval>
    | undefined;
  private exchangeInfo: ExchangeInfoResponse | undefined;
  private averageAssetPrices = new Map<string, number>();
  private tradingPairsToMonitor = new Set<string>();
  private logger: Logger;
  private apiKey: string;
  private apiSecret: string;

  constructor({
    logger,
    apiKey,
    apiSecret,
  }: {
    logger: Logger;
    apiKey: string;
    apiSecret: string;
  }) {
    super();
    this.logger = logger;
    this.apiKey = apiKey;
    this.apiSecret = apiSecret;
  }

  public start = async (activeTradingPairs?: Set<string>): Promise<any> => {
    this.pingInterval = setInterval(() => {
      this.ping();
    }, PING_INTERVAL);
    this.exchangeInfoInterval = setInterval(() => {
      this.getExchangeInfo();
    }, EXCHANGE_INFO_INTERVAL);
    if (activeTradingPairs && activeTradingPairs.size > 0) {
      this.tradingPairsToMonitor = activeTradingPairs;
    }
    this.updateAveragePricesInterval = setInterval(() => {
      this.setAveragePrices();
    }, UPDATE_AVERAGE_PRICES_INTERVAL);
    await this.ping();
    await this.getExchangeInfo();
    await this.setAveragePrices();
  };

  public monitorPriceForTradingPair = async (tradingPair: string) => {
    const previousSize = this.tradingPairsToMonitor.size;
    this.tradingPairsToMonitor.add(tradingPair);
    const newSize = this.tradingPairsToMonitor.size;
    if (newSize > previousSize) {
      const avgPriceResponse = await this.getAveragePrice(tradingPair);
      this.averageAssetPrices.set(
        avgPriceResponse.sym,
        parseFloat(avgPriceResponse.price)
      );
    }
  };

  public stopMonitoringPrice = (tradingPair: string) => {
    this.tradingPairsToMonitor.delete(tradingPair);
    this.averageAssetPrices.delete(tradingPair);
  };

  public stop = async () => {
    if (this.pingInterval) {
      clearInterval(this.pingInterval);
    }
    if (this.exchangeInfoInterval) {
      clearInterval(this.exchangeInfoInterval);
    }
    if (this.updateAveragePricesInterval) {
      clearInterval(this.updateAveragePricesInterval);
    }
  };

  public ping = async (): Promise<unknown> => {
    const response = await axios.get(PING_URL);
    assert.equal(response.status, 200);
    return response.data;
  };

  public getExchangeInfo = async (): Promise<
    ExchangeInfoResponse | undefined
  > => {
    try {
      const response = await axios.get(EXCHANGE_INFO_URL);
      assert.equal(response.status, 200);
      this.exchangeInfo = response.data;
      this.logger.info('exchangeInfo updated');
      return response.data;
    } catch (e) {
      this.logger.error(`failed to fetch exchange info ${e}`);
      return;
    }
  };

  public queryOrder = async (
    queryOrderRequest: QueryOrderRequest
  ): Promise<QueryOrderResponse> => {
    const queryOrderResponse = await this.queryOrderInternal(queryOrderRequest);
    return {
      tradingPair: queryOrderResponse.symbol,
      status: queryOrderResponse.status,
    };
  };

  public cancelOrder = async (
    cancelOrderRequest: CancelOrderRequest
  ): Promise<boolean> => {
    try {
      await this.cancelOrderInternal(cancelOrderRequest);
      return true;
    } catch (e) {
      return false;
    }
  };

  private cancelOrderInternal = async (
    cancelOrder: CancelOrderRequest
  ): Promise<void> => {
    const recvWindow = 5000;
    const timestamp = new Date().getTime();
    const { tradingPair, orderId } = cancelOrder;
    const queryString = `timestamp=${timestamp}&recvWindow=${recvWindow}&symbol=${tradingPair}&origClientOrderId=${orderId}`;
    try {
      const response = await axios.delete(
        `${ORDER_URL}?${queryString}&signature=${this.signRequest(
          queryString
        )}`,
        {
          headers: {
            'X-MBX-APIKEY': this.apiKey,
          },
        }
      );
      return response.data;
    } catch (e) {
      if (e.response && e.response.data && e.response.data.msg) {
        this.logger.error(`failed to cancel order: ${e.response.data.msg}`);
      } else {
        this.logger.error(`failed to cancel order: ${e}`);
      }
      return Promise.reject(e);
    }
  };

  private queryOrderInternal = async (
    queryOrder: QueryOrderRequest
  ): Promise<QueryOrderResponseBinance> => {
    if (process.env.LIVE_TRADING) {
      const recvWindow = 5000;
      const timestamp = new Date().getTime();
      const { tradingPair, orderId } = queryOrder;
      const queryString = `timestamp=${timestamp}&recvWindow=${recvWindow}&symbol=${tradingPair}&origClientOrderId=${orderId}`;
      try {
        const response = await axios.get(
          `${ORDER_URL}?${queryString}&signature=${this.signRequest(
            queryString
          )}`,
          {
            headers: {
              'X-MBX-APIKEY': this.apiKey,
            },
          }
        );
        return response.data;
      } catch (e) {
        if (e.response && e.response.data && e.response.data.msg) {
          this.logger.error(
            `failed to query order info: ${e.response.data.msg}`
          );
        } else {
          this.logger.error(`failed to query order info ${e}`);
        }
        return Promise.reject(e);
      }
    } else {
      // Return mock filled response when live trading is disabled
      return {
        symbol: queryOrder.tradingPair,
        status: OrderStatus.Filled,
      };
    }
  };

  public verifyQuantity = async (
    baseAsset: string,
    quoteAsset: string,
    quantity: number
  ) => {
    if (!this.exchangeInfo) {
      throw new Error('exchangeInfo has not been fetched');
    }
    const tradingPair = `${baseAsset}${quoteAsset}`;
    let avgPrice = this.averageAssetPrices.get(tradingPair);
    if (!avgPrice) {
      await this.monitorPriceForTradingPair(tradingPair);
      avgPrice = this.averageAssetPrices.get(tradingPair)!;
    }
    const filters = this.exchangeInfo!.symbols.filter(sym => {
      return sym.baseAsset === baseAsset && sym.quoteAsset === quoteAsset;
    }).map(sym => sym.filters)[0];
    const minNotionalFilter = filters.filter(
      filt => filt.filterType === 'MIN_NOTIONAL'
    )[0] as MinNotionalFilter;
    const minNotional =
      avgPrice * quantity - parseFloat(minNotionalFilter.minNotional);
    if (minNotional <= 0) {
      throw new Error('MIN_NOTIONAL filter failure');
    }
    const lotSize = filters.filter(
      filt => filt.filterType === 'LOT_SIZE'
    )[0] as LotSizeFilter;
    if (
      quantity < parseFloat(lotSize.minQty) ||
      quantity > parseFloat(lotSize.maxQty)
    ) {
      throw new Error('LOT_SIZE filter failure');
    }
    const leftOvers =
      (quantity - parseFloat(lotSize.minQty)) % parseFloat(lotSize.stepSize);
    let resultQty = quantity;
    if (leftOvers) {
      resultQty = resultQty - leftOvers;
    }
    return parseFloat(resultQty.toFixed(8));
  };

  public getAveragePrice = async (sym: string): Promise<AvgPriceResponse> => {
    const response = await axios.get(`${AVERAGE_PRICE}?symbol=${sym}`);
    assert.equal(response.status, 200);
    return { sym, price: response.data.price };
  };

  public accountInfo = async (): Promise<AccountInfo> => {
    const recvWindow = 5000;
    const timestamp = new Date().getTime();
    const queryString = `timestamp=${timestamp}&recvWindow=${recvWindow}`;
    try {
      const response = await axios.get(
        `${ACCOUNT_URL}?${queryString}&signature=${this.signRequest(
          queryString
        )}`,
        {
          headers: {
            'X-MBX-APIKEY': this.apiKey,
          },
        }
      );
      return response.data;
    } catch (e) {
      this.logger.error(`failed to fetch account info ${JSON.stringify(e)}`);
      return { balances: [] };
    }
  };

  public getAssets = async (): Promise<Balance[]> => {
    const accountInfo = await this.accountInfo();
    const parsedBalances = accountInfo.balances.map(balance => {
      return {
        asset: balance.asset,
        free: parseFloat(balance.free),
        locked: parseFloat(balance.locked),
      };
    });
    return parsedBalances;
  };

  public newOrder = (
    orderRequest: LimitOrderRequest | StopLimitOrderRequest
  ): BinanceOrder => {
    const orderId = uuidv4();
    const order = new BinanceOrder({
      ...orderRequest,
      orderId,
      api: this,
      logger: this.logger,
    });
    return order;
  };

  public startOrder = async (
    order: LimitOrderRequest | StopLimitOrderRequest
  ): Promise<string> => {
    const orderResponse = await this.newOrderInternal(order);
    if (process.env.LIVE_TRADING) {
      return orderResponse.clientOrderId;
    } else {
      return uuidv4();
    }
  };

  private newOrderInternal = async (
    order: LimitOrderRequest | StopLimitOrderRequest
  ): Promise<OrderResponse> => {
    try {
      const {
        baseAsset,
        quoteAsset,
        orderSide: side,
        orderType: type,
        quantity,
      } = order;
      const qty = await this.verifyQuantity(baseAsset, quoteAsset, quantity);
      const timestamp = new Date().getTime();
      const tradeData: { [key: string]: string | number } = {
        timestamp,
        side,
        type,
        symbol: `${baseAsset}${quoteAsset}`,
        quantity: qty,
      };
      if (type === OrderType.StopLimit || type === OrderType.Limit) {
        const { price } = order;
        if (!price) {
          return Promise.reject(
            'price is required for limit and stop-limit orders'
          );
        } else {
          tradeData.price = price;
        }
        // Currently all limit and stop-limit
        // orders are good till canceled
        tradeData.timeInForce = 'GTC';
      }
      if (type === OrderType.StopLimit) {
        const { stopPrice } = order as StopLimitOrderRequest;
        if (!stopPrice) {
          return Promise.reject('stopPrice is required for stop-limit orders');
        } else {
          tradeData.stopPrice = stopPrice;
        }
      }
      const queryStringData = querystring.stringify(tradeData);
      let url = `${TEST_ORDER_URL}?signature=${this.signRequest(
        queryStringData
      )}`;
      if (process.env.LIVE_TRADING) {
        url = `${ORDER_URL}?signature=${this.signRequest(queryStringData)}`;
      }
      const response = await axios({
        url,
        method: 'post',
        data: queryStringData,
        headers: {
          'X-MBX-APIKEY': this.apiKey,
          'Content-type': 'application/x-www-form-urlencoded',
        },
      });
      return response.data;
    } catch (e) {
      if (e.response && e.response.data && e.response.data.msg) {
        this.logger.error(`failed to create new order: ${e.response.data.msg}`);
      } else {
        this.logger.error(`failed to create new order ${e}`);
      }
      return Promise.reject(e);
    }
  };

  private setAveragePrices = async () => {
    const averagePricesPromises: Promise<any>[] = [];
    this.tradingPairsToMonitor.forEach(tp => {
      averagePricesPromises.push(this.getAveragePrice(tp));
    });
    const averagePriceResponses = await Promise.all(averagePricesPromises);
    averagePriceResponses.forEach(averagePriceResponse => {
      this.logger.info(
        `updated 5 minute average price for ${averagePriceResponse.sym}: ${averagePriceResponse.price}`
      );
      this.averageAssetPrices.set(
        averagePriceResponse.sym,
        parseFloat(averagePriceResponse.price)
      );
    });
  };

  private signRequest = (query: string): string => {
    return crypto
      .createHmac('sha256', this.apiSecret)
      .update(query)
      .digest('hex');
  };
}

export { BinanceAPI };
