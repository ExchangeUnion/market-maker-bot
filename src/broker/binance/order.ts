import {
  OrderType,
  OrderStatus,
} from '../../enums';
import {
  LimitOrderRequest,
  StopLimitOrderRequest,
} from '../exchange';
import { BinanceAPI } from './api';
import { Order } from '../order';

class BinanceOrder extends Order {
  private checkOrderInterval!: ReturnType<typeof setInterval>;

  public start = async () => {
    if (typeof this.price === 'string') {
      throw new Error('Price cannot be string for Binance orders');
    }
    const orderRequest: LimitOrderRequest | StopLimitOrderRequest = {
      orderId: this.orderId,
      baseAsset: this.baseAsset,
      quoteAsset: this.quoteAsset,
      orderSide: this.orderSide,
      orderType: this.orderType,
      quantity: this.quantity,
      price: this.price,
    };
    if (this.orderType === OrderType.StopLimit && this.stopPrice) {
      (orderRequest as StopLimitOrderRequest).stopPrice = this.stopPrice;
    }
    this.logger.info(`starting new order: ${JSON.stringify(orderRequest)}`);
    try {
      this.orderId = await (this.api as BinanceAPI).startOrder(orderRequest);
      this.logger.info(`order acknowledged with id: ${this.orderId}`);
      this.checkOrderInterval = setInterval(() => {
        this.checkOrder();
      }, 1000);
    } catch (e) {
      this.logger.info(`failed to start the order: ${JSON.stringify(orderRequest)}`);
      this.emit('failure', e);
    }
  }

  public cancel = async () => {
    if (!this.orderId) {
      return;
    }
    if (
      this.status === OrderStatus.New ||
      this.status === OrderStatus.PartiallyFilled
    ) {
      const cancelSuccess = await this.api.cancelOrder({
        tradingPair: this.tradingPair,
        orderId: this.orderId,
      });
      if (cancelSuccess) {
        this.logger.info(`canceled order with id: ${this.orderId}`);
      }
      this.status = OrderStatus.Canceled;
      this.emit('status', OrderStatus.Canceled);
      clearInterval(this.checkOrderInterval);
    }
  }

  private checkOrder = async () => {
    if (
      !this.orderId ||
      this.status === OrderStatus.Filled
    ) {
      return;
    }
    try {
      const orderInfo = await this.api.queryOrder({
        tradingPair: this.tradingPair,
        orderId: this.orderId,
      });
      if (orderInfo.status !== this.status) {
        this.status = orderInfo.status;
        this.emit('status', this.status);
      }
      if (orderInfo.status === OrderStatus.Filled) {
        // TODO: handle partial fills. Skipping for now since we're
        // switching to CCXT.
        this.emit('complete', this.orderId, this.quantity);
        clearInterval(this.checkOrderInterval);
      }
    } catch (e) {
      clearInterval(this.checkOrderInterval);
      this.emit('failure', e);
    }
  }

}

export { BinanceOrder };
