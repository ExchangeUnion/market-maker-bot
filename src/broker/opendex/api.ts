import {
  ExchangeAPI,
  Balance,
  BalanceProperties,
  CancelOrderRequest,
  QueryOrderResponse,
} from '../api';
import { Logger } from '../../logger';
import { XudGrpcClient } from './xud-client';
import { v4 as uuidv4 } from 'uuid';
import {
  OrderSide as XudOrderSide,
  PlaceOrderResponse,
  SwapSuccess,
 } from './proto/xudrpc_pb';
import { ClientReadableStream, status } from '@grpc/grpc-js';
import { OrderSide } from '../../enums';
import {
  MarketOrderRequest,
  LimitOrderRequest,
  StopLimitOrderRequest,
} from '../exchange';
import { OpenDexOrder } from './order';
import { satsToCoinsStr, coinsToSats } from '../../utils';

class OpenDexAPI extends ExchangeAPI {
  private logger: Logger;
  private certPath: string;
  private rpchost: string;
  private rpcport: number;
  private xudClient!: XudGrpcClient;
  private swapSubscriptions = new Map<string, (swapSuccess: SwapSuccess.AsObject) => unknown >();
  private checkConnectionTimeout: ReturnType<typeof setTimeout> | undefined;

  private swapsCompleteSubscription: ClientReadableStream<SwapSuccess> | undefined;

  constructor(
    { logger, certPath, rpchost, rpcport }:
    {
      logger: Logger,
      certPath: string,
      rpchost: string,
      rpcport: number,
    },
  ) {
    super();
    this.logger = logger;
    this.certPath = certPath;
    this.rpchost = rpchost;
    this.rpcport = rpcport;
  }

  public start = async () => {
    this.xudClient = new XudGrpcClient({
      tlscertpath: this.certPath,
      rpchost: this.rpchost,
      rpcport: this.rpcport,
    });
    this.xudClient.start();
    await this.waitForConnection();
    this.swapsCompleteSubscription = await this.xudClient.subscribeSwaps();
    this.swapsCompleteSubscription.on('data', this.onSwapComplete.bind(this));
    this.swapsCompleteSubscription.on('end', () => {
      this.start();
      // TODO: cancel all orders on other exchanges
    });
    this.swapsCompleteSubscription.on('error', (error: any) => {
      if (error.code && error.code === status.CANCELLED) {
        return;
      } else {
        this.logger.error(`failed to subscribe to xud swaps: ${error}`);
      }
    });
    this.logger.info('OpenDEX API started');
  }

  private waitForConnection = () => {
    return new Promise((resolve, reject) => {
      const verifyConnection = async () => {
        try {
          await this.xudClient.getBalance();
          resolve();
        } catch (e) {
          if (e.code === 14) {
            this.logger.warn('could not verify connection to xud, retrying in 5 sec...');
            this.checkConnectionTimeout = setTimeout(verifyConnection, 5000);
          } else {
            reject();
          }
        }
      };
      verifyConnection();
    });
  }

  public stop = async () => {
    if (this.checkConnectionTimeout) {
      clearTimeout(this.checkConnectionTimeout);
    }
    if (this.swapsCompleteSubscription) {
      this.swapsCompleteSubscription.cancel();
    }
  }

  public getAssets = async (): Promise<Balance[]> => {
    const getBalanceResponse = await this.xudClient.getBalance();
    const assetsMap = new Map<string, BalanceProperties>();
    getBalanceResponse.balancesMap.forEach((balance) => {
      const asset = balance[0];
      const free = parseFloat(
        satsToCoinsStr(balance[1].channelBalance),
      );
      const locked = parseFloat(
        satsToCoinsStr(balance[1].walletBalance),
      );
      assetsMap.set(asset, {
        free,
        locked,
      });
    });
    const tradingLimitsResponse = await this.xudClient.tradingLimits();
    tradingLimitsResponse.limitsMap.forEach((limit) => {
      const asset = limit[0];
      const maxbuy = limit[1].maxbuy;
      const maxsell = limit[1].maxsell;
      let balance = assetsMap.get(asset);
      if (balance) {
        balance = {
          maxbuy,
          maxsell,
          ...balance,
        };
        assetsMap.set(asset, balance);
      }
    });
    const balances: Balance[] = [];
    assetsMap.forEach((balance, asset) => {
      balances.push({
        asset,
        ...balance,
      });
    });
    return balances;
  }

  public newOrder = (
    orderRequest: MarketOrderRequest | LimitOrderRequest | StopLimitOrderRequest,
  ): OpenDexOrder => {
    const orderId = uuidv4();
    const order = new OpenDexOrder({
      ...orderRequest,
      orderId,
      api: this,
      logger: this.logger,
    });
    return order;
  }

  public startOrder = async (
    orderRequest: MarketOrderRequest | LimitOrderRequest | StopLimitOrderRequest,
  ): Promise<PlaceOrderResponse.AsObject> => {
    const pairId = `${orderRequest.baseAsset}/${orderRequest.quoteAsset}`;
    const quantityInSatoshis = coinsToSats(orderRequest.quantity);
    const xudOrderSide = orderRequest.orderSide === OrderSide.Buy
      ? XudOrderSide.BUY
      : XudOrderSide.SELL;
    const xudOrderRequest = {
      pairId,
      orderId: orderRequest.orderId,
      price: orderRequest.price,
      quantity: quantityInSatoshis,
      orderSide: xudOrderSide,
    };
    const placeOrderResponse = await this.xudClient.newOrder(xudOrderRequest);
    return placeOrderResponse;
  }

  public cancelOrder = async (cancelOrderRequest: CancelOrderRequest): Promise<boolean> => {
    try {
      await this.xudClient.removeOrder(cancelOrderRequest.orderId);
      return true;
    } catch (e) {
      return false;
    }
  }

  public subscribeSwap = (id: string, cb: (swapSuccess: SwapSuccess.AsObject) => unknown) => {
    this.swapSubscriptions.set(id, cb);
  }

  public queryOrder = async (): Promise<QueryOrderResponse> => {
    throw new Error('not implemented');
  }

  private onSwapComplete = (swapSuccess: SwapSuccess) => {
    const swapSuccessObject = swapSuccess.toObject();
    const swapSubscription = this.swapSubscriptions.get(swapSuccessObject.localId);
    swapSubscription && swapSubscription(swapSuccessObject);
  }

}

export { OpenDexAPI };
