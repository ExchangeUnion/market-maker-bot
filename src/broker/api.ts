import { OrderSide, OrderType, OrderStatus } from '../enums';
import {
  MarketOrderRequest,
  LimitOrderRequest,
  StopLimitOrderRequest,
} from './exchange';
import { BinanceOrder } from './binance/order';
import { OpenDexOrder } from './opendex/order';

export type BalanceProperties = {
  free: number;
  locked: number;
  maxsell?: number;
  maxbuy?: number;
};

export type Balance = BalanceProperties & {
  asset: string;
};

export type OrderRequest = {
  baseAsset: string;
  quoteAsset: string;
  side: OrderSide;
  type: OrderType;
  quantity: number;
  price?: number;
  stopPrice?: number;
};

export type QueryOrderResponse = {
  tradingPair: string;
  status: OrderStatus;
};

export type QueryOrderRequest = {
  tradingPair: string;
  orderId: string;
};

export type TradingLimits = {
  asset: string;
  maxsell: number;
  maxbuy: number;
};

export type CancelOrderRequest = QueryOrderRequest;

abstract class ExchangeAPI {
  public abstract start(): Promise<void>;
  public abstract stop(): Promise<void>;
  public abstract getAssets(): Promise<Balance[]>;
  public abstract newOrder(
    orderRequest: MarketOrderRequest | LimitOrderRequest | StopLimitOrderRequest
  ): BinanceOrder | OpenDexOrder;
  public abstract queryOrder(
    queryOrderRequest: QueryOrderRequest
  ): Promise<QueryOrderResponse>;
  public abstract cancelOrder(
    cancelOrderRequest: CancelOrderRequest
  ): Promise<boolean>;
}

export { ExchangeAPI };
