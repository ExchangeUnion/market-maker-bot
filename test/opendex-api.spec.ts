import { OpenDexAPI } from '../src/broker/opendex/api';
import { OpenDexOrder } from '../src/broker/opendex/order';
import { Level, Logger } from '../src/logger';
import { XudGrpcClient } from '../src/broker/opendex/xud-client';
import {
  getBalanceResponse,
  tradingLimitsResponse,
} from '../mock-data/xud-grpc-responses';
import {
  OrderType,
  OrderSide,
} from '../src/enums';
import {
  OrderSide as XudOrderSide,
  SwapSuccess,
} from '../src/broker/opendex/proto/xudrpc_pb';

jest.mock('../src/broker/opendex/xud-client');
const mockedXudGrpcClient = <jest.Mock<XudGrpcClient>><any>XudGrpcClient;
const loggers = Logger.createLoggers(Level.Warn);

describe('OpenDexAPI', () => {
  let openDexAPI: OpenDexAPI;
  let onSwapsSubscription: any;
  let onCancelSubscribeSwaps: any;

  beforeEach(async () => {
    mockedXudGrpcClient.prototype.start = jest.fn();
    onSwapsSubscription = jest.fn();
    onCancelSubscribeSwaps = jest.fn();
    mockedXudGrpcClient.prototype.subscribeSwaps = jest.fn()
      .mockImplementation(() => {
        return Promise.resolve({
          cancel: onCancelSubscribeSwaps,
          on: onSwapsSubscription,
        });
      });
    openDexAPI = new OpenDexAPI({
      logger: loggers.opendex,
      certPath: '/path/to/cert',
      rpchost: 'localhost',
      rpcport: 8886,
    });
    openDexAPI['waitForConnection'] = jest.fn()
      .mockReturnValue(Promise.resolve());
    await openDexAPI.start();
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  test('start and stop', async () => {
    expect(mockedXudGrpcClient).toHaveBeenCalledTimes(1);
    expect(mockedXudGrpcClient.prototype.start).toHaveBeenCalledTimes(1);
    expect(mockedXudGrpcClient.prototype.subscribeSwaps).toHaveBeenCalledTimes(1);
    expect(openDexAPI['xudClient']).toBeTruthy();
    expect(openDexAPI['swapsCompleteSubscription']).toBeTruthy();
    expect(onSwapsSubscription).toHaveBeenCalledTimes(3);
    expect(onSwapsSubscription).toHaveBeenCalledWith('data', expect.any(Function));
    expect(onSwapsSubscription).toHaveBeenCalledWith('error', expect.any(Function));
    expect(onSwapsSubscription).toHaveBeenCalledWith('end', expect.any(Function));
    openDexAPI.stop();
    expect(onCancelSubscribeSwaps).toHaveBeenCalledTimes(1);
  });

  test('getAssets', async () => {
    mockedXudGrpcClient.prototype.getBalance = jest.fn()
      .mockReturnValue(getBalanceResponse);
    mockedXudGrpcClient.prototype.tradingLimits = jest.fn()
      .mockReturnValue(tradingLimitsResponse);
    const ownedAssets = await openDexAPI.getAssets();
    expect(ownedAssets).toMatchSnapshot();
  });

  describe('subscribeSwap', () => {

    test('onSwapComplete does trigger the callbacks', () => {
      const cb = jest.fn();
      const cb2 = jest.fn();
      const id = '123abc';
      const id2 = '321abc';
      openDexAPI.subscribeSwap(id, cb);
      openDexAPI.subscribeSwap(id2, cb2);
      expect(openDexAPI['swapSubscriptions'].has(id)).toBeTruthy();
      const mockSwapSuccess = {
        toObject: () => {
          return {
            localId: id,
          };
        },
      };
      openDexAPI['onSwapComplete'](mockSwapSuccess as SwapSuccess);
      expect(cb).toHaveBeenCalledTimes(1);
      expect(cb2).toHaveBeenCalledTimes(0);
      const mockSwapSuccess2 = {
        toObject: () => {
          return {
            localId: id2,
          };
        },
      };
      jest.resetAllMocks();
      openDexAPI['onSwapComplete'](mockSwapSuccess2 as SwapSuccess);
      expect(cb).toHaveBeenCalledTimes(0);
      expect(cb2).toHaveBeenCalledTimes(1);
    });

  });

  describe('cancel', () => {

    test('returns true when request succeeds', async () => {
      mockedXudGrpcClient.prototype.removeOrder = jest.fn()
        .mockReturnValue(Promise.resolve(true));
      await expect(
        openDexAPI.cancelOrder(
          {
            tradingPair: 'BTC/DAI',
            orderId: '123',
          },
        )).resolves.toBeTruthy();
    });

    test('returns false when request fails', async () => {
      mockedXudGrpcClient.prototype.removeOrder = jest.fn()
        .mockReturnValue(Promise.reject(false));
      await expect(
        openDexAPI.cancelOrder(
          {
            tradingPair: 'BTC/DAI',
            orderId: '123',
          },
        )).resolves.toBeFalsy();
    });

  });

  describe('newOrder', () => {

    test('limit order succeeds', async () => {
      mockedXudGrpcClient.prototype.newOrder =
        jest.fn();
      const orderRequest = {
        orderId: '123',
        baseAsset: 'LTC',
        quoteAsset: 'BTC',
        orderType: OrderType.Limit,
        orderSide: OrderSide.Buy,
        quantity: 0.005,
        price: 0.021185,
      };
      const order = openDexAPI.newOrder(orderRequest);
      expect(order instanceof OpenDexOrder).toBeTruthy();
      await order.start();
      expect(mockedXudGrpcClient.prototype.newOrder)
        .toHaveBeenCalledTimes(1);
      expect(mockedXudGrpcClient.prototype.newOrder)
        .toHaveBeenCalledWith(
          expect.objectContaining({
            orderId: expect.any(String),
            orderSide: XudOrderSide.BUY,
            pairId: `${orderRequest.baseAsset}/${orderRequest.quoteAsset}`,
            price: orderRequest.price,
            quantity: orderRequest.quantity * 10 ** 8,
          }),
        );
    });

    test('market order succeeds', async () => {
      mockedXudGrpcClient.prototype.newOrder =
        jest.fn();
      const orderRequest = {
        orderId: '123',
        baseAsset: 'LTC',
        quoteAsset: 'BTC',
        orderType: OrderType.Market,
        orderSide: OrderSide.Sell,
        quantity: 0.005,
        price: 'mkt',
      };
      const order = openDexAPI.newOrder(orderRequest);
      expect(order instanceof OpenDexOrder).toBeTruthy();
      await order.start();
      expect(mockedXudGrpcClient.prototype.newOrder)
        .toHaveBeenCalledTimes(1);
      expect(mockedXudGrpcClient.prototype.newOrder)
        .toHaveBeenCalledWith(
          expect.objectContaining({
            orderId: expect.any(String),
            orderSide: XudOrderSide.SELL,
            pairId: `${orderRequest.baseAsset}/${orderRequest.quoteAsset}`,
            price: 'mkt',
            quantity: orderRequest.quantity * 10 ** 8,
          }),
        );
    });

  });

});
