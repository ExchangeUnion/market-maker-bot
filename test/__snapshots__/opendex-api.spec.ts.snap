// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpenDexAPI getAssets 1`] = `
Array [
  Object {
    "asset": "BTC",
    "free": 0.00116486,
    "locked": 0,
    "maxbuy": 58243,
    "maxsell": 58243,
  },
  Object {
    "asset": "DAI",
    "free": 0,
    "locked": 0,
    "maxbuy": 0,
    "maxsell": 0,
  },
  Object {
    "asset": "LTC",
    "free": 0.16579741,
    "locked": 0,
    "maxbuy": 11053160,
    "maxsell": 5526580,
  },
  Object {
    "asset": "WETH",
    "free": 0,
    "locked": 0,
    "maxbuy": 0,
    "maxsell": 0,
  },
]
`;
